TUTORIAL.adoc

```diff
@@ -0,0 +1,325 @@
+= tcell tutorial
+
+tcell provides a low-level, portable API for building terminal-based programs.
+A https://en.wikipedia.org/wiki/Terminal_emulator[terminal emulator] is used to
+interact with such a program.
+
+Applications typically initialize a screen and enter an event loop, then
+finalize the screen before exiting.
+
+Application frameworks such as https://github.com/rivo/tview[tview] and
+https://gitlab.com/tslocum/cview[cview] provide widgets and additional features.
+
+== Resize events
+
+Applications receive an event of type `EventResize` when they are first initialized and each time the terminal is resized.
+The new size is available as `Size`.
+
+[source,go]
+----
+switch ev := ev.(type) {
+case *tcell.EventResize:
+	w, h := ev.Size()
+	logMessage(fmt.Sprintf("Resized to %dx%d", w, h))
+}
+----
+
+== Key events
+
+When a key is pressed, applications receive an event of type `EventKey`.
+This event describes the modifier keys pressed (if any) and the pressed key or rune.
+
+When a rune key is pressed, an event with its `Key` set to `KeyRune` is dispatched.
+
+When a non-rune key is pressed, it is available as the `Key` of the event.
+
+[source,go]
+----
+switch ev := ev.(type) {
+case *tcell.EventKey:
+    mod, key, ch := ev.Mod(), ev.Key(), ev.Rune()
+    logMessage(fmt.Sprintf("EventKey Modifiers: %d Key: %d Rune: %d", mod, key, ch))
+}
+----
+
+=== Key event restrictions
+
+Terminal-based programs have less visibility into keyboard activity than graphical applications.
+
+When a key is pressed and held, additional key press events are sent by the terminal emulator.
+The rate of these repeated events depends on the emulator's configuration.
+Key release events are not available.
+
+It is not possible to distinguish runes typed while holding shift and runes typed using caps lock.
+Capital letters are reported without the Shift modifier.
+
+=== Key event handling library
+
+https://gitlab.com/tslocum/cbind[cbind] provides key event encoding and decoding
+to and from human-readable strings. It also provides keybinding-based input handling.
+
+== Mouse events
+
+Applications receive an event of type `EventMouse` when the mouse moves, or a mouse button is pressed or released.
+Mouse events are only delivered if
+`EnableMouse` has been called.
+
+The mouse buttons being pressed (if any) are available as `Buttons`, and the position of the mouse is available as `Position`.
+
+[source,go]
+----
+switch ev := ev.(type) {
+case *tcell.EventMouse:
+	mod := ev.Modifiers()
+	btns := ev.Buttons()
+	x, y := ev.Position()
+	logMessage(fmt.Sprintf("EventMouse Modifiers: %d Buttons: %d Position: %d,%d", mod, btns, x, y))
+}
+----
+
+=== Mouse buttons
+
+[cols=3*,options=header]
+|===
+
+|Identifier
+|Alias
+|Description
+
+|Button1
+|ButtonPrimary
+|Left button
+
+|Button2
+|ButtonSecondary
+|Right button
+
+|Button3
+|ButtonMiddle
+|Middle button
+
+|Button4
+|
+|Side button (thumb/next)
+
+|Button5
+|
+|Side button (thumb/prev)
```

Comment:
Would be nice to record that there are also WheelUp, WheelDown, and possibly WheelLeft and WheelRight events.